File: server/src/main/java/server/models/Document.java
package server.models;

import jakarta.persistence.*;
import lombok.*;
import server.models.accountant.fund.Fund;
import server.models.enums.DocumentStatus;
import server.models.enums.DocumentType;
import server.models.enums.ProjectPriority;

import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "documents")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Document {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(unique = true, nullable = false, length = 32)
    private String code;

    private String title;

    private String content;
    @Column(name = "code", length = 50, unique = true, nullable = false)
    private String code;
    private String fileUrl;

    @Column(name = "project_name")
    private String projectName;

    @Column(name = "project_description", columnDefinition = "TEXT")
    private String projectDescription;

    @Column(name = "project_deadline")
    private LocalDate projectDeadline;

    @Enumerated(EnumType.STRING)
    @Column(name = "project_priority")
    private ProjectPriority projectPriority;

    @ManyToOne
    @JoinColumn(name = "created_by_id")
    private Account createdBy;

    @ManyToOne
    @JoinColumn(name = "receiver_id")
    private Account receiver;

    @ManyToOne
    @JoinColumn(name = "pm_id")
    private Account pm;

    @ManyToOne
    @JoinColumn(name = "accountant_id")
    private Account accountant;

    @OneToOne(mappedBy = "document")
    private Project project;

    @OneToOne(mappedBy = "document")
    private Fund relatedFund;

    @Column(name = "fund_name")
    private String fundName;

    @Column(name = "fund_balance")
    private Double fundBalance;

    @Column(name = "fund_purpose", columnDefinition = "TEXT")
    private String fundPurpose;

    private LocalDateTime createdAt;

    @Enumerated(EnumType.STRING)
    private DocumentStatus status;

    @Enumerated(EnumType.STRING)
    private DocumentType type;

    @Column(columnDefinition = "TEXT")
    private String signature;
}

File: mobile/lib/models/document.dart
import 'package:mobile/models/enums/document_status.dart';
import 'package:mobile/models/enums/document_type.dart';

class DocumentModel {
  final int id;
  final String code;
  final String title;
  final String content;
  final String? fileUrl;
  final String createdBy;
  final String? receiver;
  final int? relatedProjectId;
  final DateTime createdAt;
  final DocumentStatus status;
  final DocumentType type;
  final String? signature;
  final String? previewHtml;

  // 📝 Thêm trường ghi chú của giám đốc
  final String? managerNote;

  // Project fields
  final String? projectName;
  final String? projectDescription;
  final String? projectPriority;
  final String? projectDeadline;
  final String? pmName;
  final int? pmId;

  // Fund fields
  final String? fundName;
  final double? fundBalance;
  final String? fundPurpose;
  final String? accountantName;

  DocumentModel({
    required this.id,
    required this.code,
    required this.title,
    required this.content,
    required this.fileUrl,
    required this.createdBy,
    required this.receiver,
    required this.relatedProjectId,
    required this.createdAt,
    required this.status,
    required this.type,
    required this.signature,
    required this.previewHtml,
    this.managerNote,
    this.projectName,
    this.projectDescription,
    this.projectPriority,
    this.projectDeadline,
    this.pmName,
    this.pmId,
    this.fundName,
    this.fundBalance,
    this.fundPurpose,
    this.accountantName,
  });

  factory DocumentModel.fromJson(Map<String, dynamic> json) {

    double? _toDouble(dynamic v) {
      if (v == null) return null;
      if (v is num) return v.toDouble();
      return double.tryParse(v.toString());
    }

    int? _toInt(dynamic v) {
      if (v == null) return null;
      if (v is num) return v.toInt();
      return int.tryParse(v.toString());
    }

    return DocumentModel(
      id: _toInt(json['id']) ?? 0,
      code: json['code']?.toString() ?? '',
      title: json['title']?.toString() ?? '',
      content: json['content']?.toString() ?? '',
      fileUrl: json['fileUrl']?.toString(),
      createdBy: json['createdBy']?.toString() ?? '',
      receiver: json['receiver']?.toString(),
      relatedProjectId: _toInt(json['relatedProjectId']),
      createdAt: DateTime.tryParse(json['createdAt']?.toString() ?? '') ?? DateTime.now(),
      status: DocumentStatusX.fromString(json['status']),
      type: DocumentTypeX.fromString(json['type']),
      signature: json['signature']?.toString(),
      previewHtml: json['previewHtml']?.toString(),

      managerNote: json['managerNote']?.toString(),

      projectName: json['projectName']?.toString(),
      projectDescription: json['projectDescription']?.toString(),
      projectPriority: json['projectPriority']?.toString(),
      projectDeadline: json['projectDeadline']?.toString(),
      pmName: json['pmName']?.toString(),
      pmId: _toInt(json['pmId']),

      fundName: json['fundName']?.toString(),
      fundBalance: _toDouble(json['fundBalance']),
      fundPurpose: json['fundPurpose']?.toString(),
      accountantName: json['accountantName']?.toString(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'code': code,
      'title': title,
      'content': content,
      'fileUrl': fileUrl,
      'createdBy': createdBy,
      'receiver': receiver,
      'relatedProjectId': relatedProjectId,
      'createdAt': createdAt.toIso8601String(),
      'status': status.name,
      'type': type.name,
      'signature': signature,
      'previewHtml': previewHtml,

      'managerNote': managerNote,

      'projectName': projectName,
      'projectDescription': projectDescription,
      'projectPriority': projectPriority,
      'projectDeadline': projectDeadline,
      'pmId': pmId,

      'fundName': fundName,
      'fundBalance': fundBalance,
      'fundPurpose': fundPurpose,
      'accountantName': accountantName,
    };
  }

}
